### ISSUE TITLE
Tildagon vs Flopagon! Invalid magic header

### ISSUE THREAD

**Issue by andypiper** [hairymnstr/Flopagon#1]
when the Tildagon starts, it reports:

```
Hexpansion inserted in port: 1
Failed to decode header: Invalid magic in hexpansion header: b'\xff\xff\xff\xff'
```

I'm guessing this is something to do with the EEPROM contents, and probably not an easy fix.

**Comment by andypiper** [comment]
Actually, presumably there's a version of [the incantations listed here](https://github.com/walkerdanny/caffeine-jitters?tab=readme-ov-file#note) that [sh|w]ould enable the intended functionality?

**Comment by andypiper** [comment]
(I'm thinking, personal webserver with the whole of wikipedia, on flopagon...)

**Comment by andypiper** [comment]
(walking NAS, with multiple flopagons...)

**Comment by hairymnstr** [comment]
Yeah "Invalid magic" means the EEPROM doesn't have a valid header.  It's because the EEPROMs on the Flopagon's I distributed aren't provisioned.  Mostly this is because there are some issues with the formatting for littlefs on the EEPROM I used.  There's a branch https://github.com/emfcamp/badge-2024-software/tree/eeprom_stuff which has some fixes on it but I still haven't had time to test that branch against other EEPROMs to make sure I'm not breaking other people's hexpansions by fixing this one.

The invalid magic error shouldn't affect your ability to use the Flash memory chip on the hexpansion.

The plan for the Flopagon EEPROM was a tiny app that's in this repo that just helps you format/mount/unmount the Flash but I think the more interesting option is a chain loader that triggers an app from the flash so you can launch into your offline Wikipedia, hitch hikers guide or whatever.

**Comment by Robotmad** [comment]
Nathan - we had the same issues with programming EEPROM using the badge software - but have now written a robust app which works with our 64kbit device (it uses 2 address bytes) - using minimal direct I2C access. It scans all hexpansion ports on start up, responds to expansion insertion/removal and offers to program the EEPROM header and then to copy an app over (in our case renaming hexdrive.mpy to app.mpy) Its part of BadgeBot: https://github.com/TeamRobotmad/BadgeBot you are of course most welcome to use any of this code.

**Comment by hairymnstr** [comment]
Hi, thanks for the suggestion I will take a look. It sounds very much like the MD Updater that Matthew wrote for his mega drive hexpansion.

The issue isn't the provisioning tool on the badge, it's the low level driver support for the EEPROM I used (even though it's on the recommended list). I'm working to address the issues without breaking support for 16bit addressed EEPROMs here: https://github.com/emfcamp/badge-2024-software/issues/129

**Comment by Robotmad** [comment]
Great - there are lots of issues and even some PRs related to the EEPROMs, needs someone to get a grip on it sort in one go - very happy to help test if the EEPROM we've used is different to those you have access to.  

It looked to me that there was an update about a month ago which broke write_header for writing to 16-bit EEPROMs, at least ours (probably someone trying to get smaller EEPROMs working).  It split the write into two parts - a) line 101: write the address, b) line 113: write the header - but the first two bytes of the second write are taken as the address again.

Once it has been worked out if an EEPROM is 8 or 16 bit addressing I think there should be a flag remembered for this rather than working it out dynamically all over the place.  Even once you've read the header I'm not sure you can tell from the EEPROM size - isn't it possible to have say 1024bytes either as 8 bit address plus 2 bits in the chip address, or use 10 bits of a 16bit address?
cbarnes@tmsl.com if you want to agree an easier means to chat.
